### PR TITLE
Add AWS Roles module

### PR DESCRIPTION
**Pull request by oharan2**
Required by [this](https://access.redhat.com/documentation/en-us/red_hat_openshift_api_management/1/guide/53dfb804-2038-4545-b917-2cb01a09ef98#_b5f80fce-73cb-4869-aa16-763bbe09896a:~:text=In%20the%20AWS%20CLI%2C%20create%20a%20policy%20for%20SRE%20Support.%20Enter%20the%20following%3A)  doc for addon installation.

The `rhoam_add_role_policy.py` internal calls will eventually be used within the `ocm_python_wrapper/cluster.py` instead of this repository.

The client resources handling is temporary, and will be provided by a more generic solution that I've created for [this](https://issues.redhat.com/browse/MSINTEROP-326) issue.

### CODE REVIEW

**Review by rnetser — CHANGES_REQUESTED** [review]
- please create a module for aws roles and a module for aws client
- create a sub-dir for roles and create a readme for it
- for rhoam - please add ref to the docs

File: clouds/aws/aws_clients.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    def list_role_policies(self, role_name):
+        return self.client.list_role_policies(RoleName=role_name)
+
+    def put_role_policy(self, role_name, policy_name, policy_document):
+        return self.client.put_role_policy(
+            RoleName=role_name,
+            PolicyName=policy_name,
+            PolicyDocument=policy_document,
+        )
```
Comment (rnetser) [inline clouds/aws/aws_clients.py:33]: this should not go under this module; relevant to roles

File: clouds/aws/aws_roles/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+## AWS Roles
+
+This repository manages IAM roles and role-policies with [Boto3 IAM client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html).
+
+
+An `IAM role` is an IAM identity that you can create in your account that has specific permissions.
+
+We will create a `role policy` in AWS by creating policies and attaching them to IAM identities.
```
Comment (rnetser) [inline clouds/aws/aws_roles/README.md:8]:
```suggestion
Create a `role policy` by creating policies and attaching them to IAM identities.
```

File: clouds/aws/aws_roles/README.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+`IAM policies` define permissions for an action regardless of the method that you use to perform the operation.
+Most policies are stored in AWS as JSON documents
```
Comment (rnetser) [inline clouds/aws/aws_roles/README.md:11]: please remove; the README is about the repo, not AWS

File: clouds/aws/aws_roles/aws_roles.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+
+class AWSRoles:
+    def __init__(self, client: IAMClient):
+        """
+        Manages AWS roles and role-policies.
+        Args:
+            client (IAMClient): AWS roles managements expects an IAM client entity
+        """
+        self.client = client
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:15]: this is not needed
get the client where needed

File: clouds/aws/aws_roles/aws_roles.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+
+class AWSRolePolicies(AWSRoles):
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:18]: no need to wrap this in a class

File: clouds/aws/aws_roles/aws_roles.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    def __init__(self, client):
+        super().__init__(client=client)
+
+    def create_role_policy(self, role_name, policy_name, policy_json_path):
+        """
+        Create the role policy with specifications.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:24]: `specifications` - what does it mean?

File: clouds/aws/aws_roles/aws_roles.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+                LOGGER.info("Done")
+            except Exception as exc:
+                LOGGER.info(f"Failed with role policy creation:\n{exc}")
+
+    def check_available_role_policy(self, role_name=""):
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:53]: `role_name` should be mandatory

File: clouds/aws/aws_roles/aws_roles.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        """
+        Finds role policy by given name, using boto3 list_roles_policies.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:55]: a docstring should not include implementation but explain what the function does

File: clouds/aws/aws_roles/aws_roles.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+        Args:
+            role_name (str): role policy name
+
+        Returns:
+            False if the request fails with botocore 'NoSuchEntity' exception.
+             This is the only way to check existence since the list method raises if the role doesn't exist.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:62]: True if a policy role is found else False.
(do not include implementation)

File: clouds/aws/aws_roles/aws_roles.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+        Args:
+            role_name (str): role policy name
+            policy_name (str): policy name
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
+        """
+        if not self.check_available_role_policy(role_name):
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:34]: use arg=val

File: clouds/aws/aws_roles/aws_roles.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+            LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+            try:
+                policy_doc = open(policy_json_path, "r").read()
+            except Exception:
+                LOGGER.info(
+                    "Json policy document couldn't load. Please validate file path."
+                )
+                raise
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:42]: use `with`

File: clouds/aws/aws_roles/aws_roles.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+            try:
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+                LOGGER.info("Done")
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:49]: not needed

File: clouds/aws/aws_roles/aws_roles.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+        Args:
+            role_name (str): role policy name
+            policy_name (str): policy name
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
+        """
+        if not self.check_available_role_policy(role_name):
+            LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:35]: should be before creation of role

File: clouds/aws/aws_roles/aws_roles.py [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+                )
+                raise
+            try:
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+                LOGGER.info("Done")
+            except Exception as exc:
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:50]: this is too broad; use explicit exception(s)

File: clouds/aws/aws_roles/aws_roles.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+                LOGGER.info(f"Failed with role policy creation:\n{exc}")
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:51]:
```suggestion
                LOGGER.error(f"Failed with role policy creation:\n{exc}")
```
should raise here

File: clouds/aws/aws_roles/rhoam_add_role_policy.py [hunk 16]
```diff
@@ -0,0 +1,20 @@
+from aws import aws_clients
```
Comment (rnetser) [inline clouds/aws/aws_roles/rhoam_add_role_policy.py:1]: The repo is for cloud tools; so rhoam-specific code should not be in this repo.
This file should be delete (and code used where needed); json file should be moved next to the new code. (also remove text from README)

Comment (oharan2) [reply clouds/aws/aws_roles/rhoam_add_role_policy.py:1]: No problem- so which exact location did [this](https://github.com/RedHatQE/cloud-tools/pull/14#:~:text=for%20rhoam%20%2D%20please%20add%20ref%20to%20the%20docs) referred to?
_for rhoam - please add ref to the docs_

File: clouds/aws/aws_clients.py [hunk 17]
```diff
@@ -0,0 +1,33 @@
+import boto3
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+class AWSClient:
+    def __init__(self, *args, **kwargs):
+        """Default region is us-east-1 also when param region_name is not given"""
+        region = kwargs.get("region") or "us-east-1"
```
[the remark this answers is not in the document]
Comment (oharan2) [reply clouds/aws/aws_clients.py]: Done

File: clouds/aws/aws_roles/README.md (hunk 2 above)
Comment (oharan2) [reply clouds/aws/aws_roles/README.md]: Done

File: clouds/aws/aws_roles/aws_roles.py (hunk 14 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/README.md (hunk 3 above)
Comment (oharan2) [reply clouds/aws/aws_roles/README.md]: Done

File: clouds/aws/aws_roles/aws_roles.py (hunk 5 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py (hunk 7 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py (hunk 10 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py (hunk 11 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+        Args:
+            role_name (str): role policy name
+            policy_name (str): policy name
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
+        """
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+        if not self.check_available_role_policy(role_name=role_name):
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:32]: Don't we want to raise if check_available_role_policy?
if line 32 not match you return None, and the user think all was done.

File: clouds/aws/aws_clients.py (hunk 1 above)
Comment (oharan2) [reply clouds/aws/aws_clients.py:33]: I can't, the role managed functions are called by class IAM.Client object (must be client.put_role_policy(..)).

At the beginning I tried to separate, but couldn't pass pre-commit with making an AWSClient (our's) object calling an IAM.Client (boto3) method (put_role_policy) without declaring it as AWSClient's method.
[Boto3 documents](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html)

File: clouds/aws/aws_roles/aws_roles.py [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+                )
+                raise
+            try:
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+            except BotoCoreError as exc:
+                LOGGER.error(f"Failed with role policy creation:\n{exc}")
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:48]: missing raise?

File: clouds/aws/aws_roles/aws_roles.py [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+
+        Raises:
+            If POST request failed
+        """
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+        if not self.check_available_role_policy(role_name=role_name):
+            try:
+                with open(policy_json_path, "r") as f:
+                    policy_doc = f.read()
+            except Exception:
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:36]: Please use specific Exceptions

File: clouds/aws/aws_roles/aws_roles.py [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+            policy_name (str): policy name
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
+        """
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+        if not self.check_available_role_policy(role_name=role_name):
+            try:
+                with open(policy_json_path, "r") as f:
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:34]: `f` > please don;t use one letter words (fd is good)

File: clouds/aws/aws_roles/aws_roles.py (hunk 4 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: ditto (must be called by client instance such as `client._aws_roles_methods_`)

File: clouds/aws/aws_roles/aws_roles.py [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+    def check_available_role_policy(self, role_name: str):
+        """
+        Finds role policy existence by given name.
+
+        Args:
+            role_name (str): role policy name
+
+        Returns:
+            True if a policy role is found else False.
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:58]: syntex is:

```
Returns:
    type: description

Returns:
            bool: True if a policy role is found else False.
```

Please fix where needed.

File: clouds/aws/aws_roles/rhoam-sre-support-policy.json [hunk 23]
```diff
@@ -0,0 +1,26 @@
+{
```
Comment (myakove) [inline clouds/aws/aws_roles/rhoam-sre-support-policy.json:1]: Do we want this file to be part of this repo?

File: clouds/aws/aws_roles/aws_roles.py [hunk 24]
```diff
@@ -0,0 +1,66 @@
...
+            try:
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+            except BotoCoreError as exc:
+                LOGGER.error(f"Failed with role policy creation:\n{exc}")
+
+    def check_available_role_policy(self, role_name: str):
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:50]: I think exists will fit here better
`check_if_role_policy_exists`

**Review by myakove — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_clients.py (hunk 1 above)
Comment (oharan2) [reply clouds/aws/aws_clients.py:33]: Would love to know how I should handle it optimally given this information.

File: clouds/aws/aws_roles/aws_roles.py (hunk 18 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: The check method logging in [line 63](https://github.com/RedHatQE/cloud-tools/pull/14/files#diff-773825122a0092f2831ee9d5485dc08a8701fe6a32a2ea7ad33c0dfb2cbbb95dR63) LOGGER.info(f"Policy role {role_name} exists.")

File: clouds/aws/aws_roles/aws_roles.py [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
+        """
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+        if not self.check_if_role_policy_exists(role_name=role_name):
+            try:
+                with open(policy_json_path, "r") as fd:
+                    policy_doc = fd.read()
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:35]: if this is a json file please use json to read it.

File: clouds/aws/aws_roles/aws_roles.py [hunk 26]
```diff
@@ -29,11 +29,11 @@ def create_role_policy(self, role_name, policy_name, policy_json_path):
...
         LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
-        if not self.check_available_role_policy(role_name=role_name):
+        if not self.check_if_role_policy_exists(role_name=role_name):
             try:
-                with open(policy_json_path, "r") as f:
-                    policy_doc = f.read()
-            except Exception:
+                with open(policy_json_path, "r") as fd:
+                    policy_doc = fd.read()
+            except (FileNotFoundError, OSError):
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:36]: `            except (FileNotFoundError, OSError):`
Check sonar cloud errors

File: clouds/aws/aws_roles/aws_roles.py (hunk 20 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done with `FileNotFoundError, OSError`

File: clouds/aws/aws_roles/aws_roles.py (hunk 21 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+
+        Args:
+            role_name (str): role policy name
+
+        Returns:
+            bool: True if a policy role is found else False.
+        """
+        try:
+            self.client.list_role_policies(role_name=role_name)
+            LOGGER.info(f"Policy role {role_name} exists.")
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:63]: what if I have role with the same name but with different set of actions?
You will use the existing role without apply the new one.

File: clouds/aws/aws_roles/aws_roles.py [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+                self.client.put_role_policy(
+                    role_name=role_name,
+                    policy_name=policy_name,
+                    policy_document=policy_doc,
+                )
+            except BotoCoreError as exc:
+                LOGGER.error(f"Failed with role policy creation:\n{exc}")
+                raise
+
+    def check_if_role_policy_exists(self, role_name: str):
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:51]: either set types for all function or not
and if you set to all of them include retrun type as well

**Review by myakove — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_roles/rhoam-sre-support-policy.json (hunk 23 above)
Comment (oharan2) [reply clouds/aws/aws_roles/rhoam-sre-support-policy.json:1]: Will delete all rhoam related code from this repo (this file and [rhoam_add_role_policy.py](https://github.com/RedHatQE/cloud-tools/blob/962bbd0eaa28ce2cfdc01358dbb28adcc5483918/clouds/aws/aws_roles/rhoam_add_role_policy.py#L13)). 
The alternative location will be within the openshift-ci scenario files?
@rnetser @ILpinto

File: clouds/aws/aws_clients.py (hunk 1 above)
Comment (rnetser) [reply clouds/aws/aws_clients.py:33]: you can delete this class and have a function in aws_roles which returns the client

File: clouds/aws/aws_roles/README.md [hunk 29]
```diff
@@ -0,0 +1,20 @@
...
+
+This repository manages IAM roles and role-policies with [Boto3 IAM client](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html).
+
+
+An `IAM role` is an IAM identity that you can create in your account that has specific permissions.
+
+Create a `role policy` by creating policies and attaching them to IAM identities.
+
+
+##  Using `rhoam_add_role_policy.py`:
```
Comment (rnetser) [inline clouds/aws/aws_roles/README.md:11]: pleaser remove all rhoam-related code

File: clouds/aws/aws_roles/aws_roles.py [hunk 30]
```diff
@@ -0,0 +1,67 @@
+import botocore.errorfactory
+from aws.aws_clients import IAMClient
+from botocore.exceptions import BotoCoreError
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+class AWSRoles:
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:9]: no need for a class here; have a function to get the client and then fucntions which get it as arg

File: clouds/aws/aws_roles/aws_roles.py [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+        """
+        Create the role policy with specific configuration.
+
+        Args:
+            role_name (str): role policy name
+            policy_name (str): policy name
+            policy_json_path (str): path to json file that holds the policy documents
+
+        Raises:
+            If POST request failed
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:29]: please update the docstring format

File: clouds/aws/aws_roles/rhoam-sre-support-policy.json (hunk 23 above)
Comment (rnetser) [reply clouds/aws/aws_roles/rhoam-sre-support-policy.json:1]: no, should be in addon isntall - (ocp-addons-operators-cli)

**Review by rnetser — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_clients.py (hunk 1 above)
Comment (oharan2) [reply clouds/aws/aws_clients.py]: Done

File: clouds/aws/aws_roles/rhoam-sre-support-policy.json (hunk 23 above)
Comment (oharan2) [reply clouds/aws/aws_roles/rhoam-sre-support-policy.json]: Removed.

File: clouds/aws/aws_roles/aws_roles.py [hunk 32]
```diff
@@ -0,0 +1,70 @@
...
+        policy_document_path (str): path to Json file that holds the policy documents
+        apply_update (bool): True to update the role policy if found, else False
+
+    Raises:
+        OSError: If file read failed
+        BotoCoreError: If POST request failed
+    """
+    client = get_iam_client()
+    LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+    if apply_update or not check_if_role_policy_exists(
```
Comment (myakove) [inline clouds/aws/aws_roles/aws_roles.py:31]: I don't think `apply_update` needed, if the user call this function and pass role he want to create/update it.

File: clouds/aws/aws_roles/aws_roles.py (hunk 25 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Tried at first, the policy update expects a string which holds **Json format content**, not a parsed (loaded) Json.
example input file:
'{ "Policy":"policy_content"}'

**Review by myakove — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_roles/aws_roles.py (hunk 26 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Replaced. Very helpful.

File: clouds/aws/aws_roles/aws_roles.py (hunk 27 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Updated with new [argument](https://github.com/RedHatQE/cloud-tools/pull/14/files#diff-773825122a0092f2831ee9d5485dc08a8701fe6a32a2ea7ad33c0dfb2cbbb95dR23) `apply_update`

File: clouds/aws/aws_roles/aws_roles.py (hunk 32 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py:31]: So we want to delete [check_if_role_policy_exists](https://github.com/RedHatQE/cloud-tools/pull/14/files#diff-773825122a0092f2831ee9d5485dc08a8701fe6a32a2ea7ad33c0dfb2cbbb95dR53)? If that is the work assumption, I don't see a reason to have it.

Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Updated  in recent commit

File: clouds/aws/aws_clients.py [hunk 33]
```diff
@@ -0,0 +1,18 @@
+import boto3
+from botocore.exceptions import BotoCoreError
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+class AWSClient:
+    def __init__(self, region="us-east-1", **kwargs):
+        """Default region is us-east-1 also when param region_name is not given"""
```
Comment (rnetser) [inline clouds/aws/aws_clients.py:10]: please use proper docsting

File: clouds/aws/aws_clients.py [hunk 34]
```diff
@@ -0,0 +1,18 @@
...
+        LOGGER.info(
+            f"Creating {kwargs.get('service_name') or 'AWS'} client using region {region}."
```
Comment (rnetser) [inline clouds/aws/aws_clients.py:12]: already commented on this - `service_name` should be an arg

File: clouds/aws/aws_roles/aws_roles.py [hunk 35]
```diff
@@ -0,0 +1,66 @@
+import botocore.errorfactory
+from aws.aws_clients import AWSClient
+from botocore.exceptions import BotoCoreError
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+def get_iam_client():
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:9]:
```suggestion
def iam_client():
```

File: clouds/aws/aws_roles/aws_roles.py [hunk 36]
```diff
@@ -0,0 +1,66 @@
...
+    return AWSClient(service_name="iam").client
+
+
+def create_role_policy(role_name, policy_name, policy_document_path):
+    """
+    Create the role policy with specific configuration.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:15]: what does `with specific configuration.` mean?

File: clouds/aws/aws_roles/aws_roles.py [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+def get_iam_client():
+    return AWSClient(service_name="iam").client
+
+
+def create_role_policy(role_name, policy_name, policy_document_path):
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:13]: the code does not `create` it `create_or_update`

File: clouds/aws/aws_roles/aws_roles.py [hunk 38]
```diff
@@ -0,0 +1,66 @@
...
+        role_name (str): role policy name
+        policy_name (str): policy name
+        policy_document_path (str): path to Json file that holds the policy documents
+
+    Raises:
+        OSError: If file read failed
+        BotoCoreError: If POST request failed
+    """
+    client = get_iam_client()
+    exists = check_if_role_policy_exists(iam_client=client, role_name=role_name)
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:27]: used once, no need to save

File: clouds/aws/aws_roles/aws_roles.py [hunk 39]
```diff
@@ -0,0 +1,66 @@
...
+    if exists:
+        LOGGER.info(f"Updating role by documented policy in {policy_document_path}.")
+    else:
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+    try:
+        with open(policy_document_path, "r") as fd:
+            policy_document = fd.read()
+    except OSError:
+        LOGGER.error("Json policy document couldn't load. Please validate file path.")
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:36]: add path to the log

File: clouds/aws/aws_roles/aws_roles.py [hunk 40]
```diff
@@ -0,0 +1,66 @@
...
+            RoleName=role_name,
+            PolicyName=policy_name,
+            PolicyDocument=policy_document,
+        )
+    except BotoCoreError as exc:
+        LOGGER.error(f"Failed with role policy management:\n{exc}")
+        raise
+
+
+def check_if_role_policy_exists(iam_client, role_name):
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:49]:
```suggestion
def role_policy_exists_by_name(iam_client, role_name):
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_clients.py (hunk 34 above)
Comment (oharan2) [reply clouds/aws/aws_clients.py:12]: Misunderstand - I want to have the default AWS client session creation, which **doesn't** get service_name argument.
[Reference -](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/boto3.html#:~:text=Create%20a%20low%2Dlevel%20service%20client%20by%20name%20using%20the%20default%20session.)
_Create a low-level service client by name using the default session._

Comment (oharan2) [reply clouds/aws/aws_clients.py:12]: For a better description passing `service_name=None` returns with error:

`Unknown service: 'None'. Valid service names are: accessanalyzer, account, acm....`

File: clouds/aws/aws_roles/aws_roles.py (hunk 36 above)
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py:15]: Required configurations such as the unique role name combined with policy name and policy doc with specified instructions. Delete?

Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py:15]: Required configurations such as the unique role name combined with policy name and policy doc with specified instructions. Delete?

File: clouds/aws/aws_roles/aws_roles.py [hunk 41]
```diff
@@ -0,0 +1,68 @@
...
+        LOGGER.info(f"Updating role by documented policy in {policy_document_path}.")
+    else:
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+    try:
+        with open(policy_document_path, "r") as fd:
+            policy_document = fd.read()
+    except OSError:
+        LOGGER.error(
+            f"Json policy document couldn't load. Please validate file path, "
+            f"got: '{policy_document_path}'."
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:37]:
```suggestion
            f"Json policy document couldn't load. Please validate file path: '{policy_document_path}'."
```

File: clouds/aws/aws_roles/aws_roles.py [hunk 42]
```diff
@@ -0,0 +1,68 @@
...
+        LOGGER.error(f"Failed with role policy management:\n{exc}")
+        raise
+
+
+def role_policy_exists_by_name(iam_client, role_name):
+    """
+    Finds role policy existence by given name.
+
+    Args:
+        iam_client (AWSClient): AWS client active session, expects an IAM client entity
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:56]:
```suggestion
        iam_client (AWSClient): AWS client
```

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by myakove — CHANGES_REQUESTED** [review]

File: clouds/aws/aws_roles/aws_roles.py [hunk 43]
```diff
@@ -0,0 +1,83 @@
...
+    """
+    client = iam_client()
+    if role_policy_exists_by_name(iam_client=client, role_name=role_name):
+        LOGGER.info(f"Updating role by documented policy in {policy_document_path}.")
+    else:
+        LOGGER.info(f"Creating new role {role_name} for policy {policy_name}.")
+    try:
+        with open(policy_document_path, "r") as fd:
+            policy_document = fd.read()
+    except OSError:
```
[the remark this answers is not in the document]
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py:50]: This case catches both `FileNotFoundError` **and** files loading/reading OS errors, do you think it's unnecessary?

File: clouds/aws/aws_roles/aws_roles.py [hunk 44]
```diff
@@ -0,0 +1,83 @@
...
+        policy_name (str): policy name
+        policy_document_path (str): path to Json file that holds the policy documents
+
+    Raises:
+        OSError: If file read failed
+        BotoCoreError: If POST request failed
+    """
+    client = iam_client()
+    if role_policy_exists_by_name(iam_client=client, role_name=role_name):
+        LOGGER.info(f"Updating role by documented policy in {policy_document_path}.")
```
[the remark this answers is not in the document]
Comment (oharan2) [reply clouds/aws/aws_roles/aws_roles.py]: Done

File: clouds/aws/aws_roles/aws_roles.py [hunk 45]
```diff
@@ -0,0 +1,85 @@
...
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+def iam_client(region="us-east-1"):
+    """Creates an IAM client.
+
+    Args:
+        region (str): Region to use for session, a client is associated with a single region. Defaults to us-east-1.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:13]:
```suggestion
        region (str, default: "us-east-1"): Region to use for session, a client is associated with a single region.
```

File: clouds/aws/aws_roles/aws_roles.py (hunk 43 above)
[the remark this answers is not in the document]
Comment (rnetser) [reply clouds/aws/aws_roles/aws_roles.py:52]: The error in the log is about file path

File: clouds/aws/aws_roles/aws_roles.py [hunk 46]
```diff
@@ -0,0 +1,85 @@
...
+
+    Returns:
+        bool: True if a policy role is found else False.
+    """
+    try:
+        iam_client.list_role_policies(RoleName=role_name)
+        LOGGER.info(f"Policy role {role_name} exists.")
+        return True
+    except botocore.errorfactory.ClientError:
+        # Fails when role policy doesn't exist.
```
Comment (rnetser) [inline clouds/aws/aws_roles/aws_roles.py:84]: please add a log

**Review by rnetser — CHANGES_REQUESTED** [review]

**Review by myakove — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Pull request merged**